File: app/src/test/java/ru/iandreyshev/parserrss/models/repository/DatabaseTest.java
package ru.iandreyshev.parserrss.models.repository;

import android.support.annotation.NonNull;

import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.*;

import java.io.File;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

public class DatabaseTest {
    private static final String RSS_TITLE = "TITLE";
    private static final String RSS_ORIGIN = "ORIGIN";
    private static final String RSS_URL = "URL";
    private static final String NOT_USE_RSS_URL = "NOT_USE_URL";

    private static final int ARTICLES_COUNT = 25;
    private static final String ARTICLE_TITLE = "TITLE %s";
    private static final String ARTICLE_DESCRIPTION = "DESCRIPTION %s";
    private static final String ARTICLE_ORIGIN = "ORIGIN %s";

    private static final byte[] IMAGE = {0, 1, 2, 3};

    private Database mDatabase;
    private Rss mRss;

    @Before
    public void setup() throws Exception {
        File tempFile = File.createTempFile("object-store-test", "");
        tempFile.delete();
        mDatabase = new Database(MyObjectBox.builder().directory(tempFile).build());

        mRss = new Rss(RSS_TITLE, RSS_ORIGIN);
        mRss.setUrl(RSS_URL);
        final List<Article> articleList = new ArrayList<>();

        for (int i = 0; i < ARTICLES_COUNT; ++i) {
            articleList.add(createArticle(i));
        }

        mRss.setArticles(articleList);
    }

    @Test
    public void noThrowExceptionIfTryToGetItemByInvalidId() {
        try {
            for (Long id : Database.INVALID_IDS) {
                mDatabase.getArticleById(id);
                mDatabase.getRssById(id);
                mDatabase.updateArticleImage(id, null);
                mDatabase.removeRssById(id);
            }
        } catch (Exception ex) {
            fail();
        }
    }

    @Test
    public void returnNullIfGetRssByInvalidId() {
        assertNull(mDatabase.getRssById(-2));
    }

    @Test
    public void returnTrueIfRssWithSameUrlExist() throws Exception {
        mDatabase.putRssIfSameUrlNotExist(mRss);

        assertTrue(mDatabase.isRssWithUrlExist(RSS_URL));
    }

    @Test
    public void returnFalseIfRssWithSameUrlNotExist() throws Exception {
        mDatabase.putRssIfSameUrlNotExist(mRss);

        assertFalse(mDatabase.isRssWithUrlExist(NOT_USE_RSS_URL));
    }

    @Test
    public void returnArticlesById() throws Exception {
        mDatabase.putRssIfSameUrlNotExist(mRss);

        for (final Article article : mRss.getArticles()) {
            assertNotNull(mDatabase.getArticleById(article.getId()));
        }
    }

    @Test
    public void returnEmptyListAfterGetAllIfStoreEmpty() throws Exception {
        assertTrue(mDatabase.getAllRss().isEmpty());
    }

    @Test
    public void updateRssIdAfterPut() throws Exception {
        mDatabase.putRssIfSameUrlNotExist(mRss);

        assertEquals(mDatabase.getRssById(mRss.getId()), mRss);
    }

    @Test
    public void returnTrueWhenUpdatingRssExist() throws Exception {
        final Rss newRss = new Rss("Title", "Origin");
        newRss.setDescription("Description");
        newRss.setUrl(mRss.getUrl());

        assertNotEquals(newRss.getTitle(), mRss.getTitle());
        assertNotEquals(newRss.getDescription(), mRss.getDescription());
        assertNotEquals(newRss.getArticles().size(), mRss.getArticles().size());

        assertEquals(newRss.getUrl(), mRss.getUrl());

        assertTrue(mDatabase.putRssIfSameUrlNotExist(newRss));
        assertTrue(mDatabase.updateRssWithSameUrl(mRss));

        assertNotEquals(newRss.getTitle(), mRss.getTitle());
        assertNotEquals(newRss.getDescription(), mRss.getDescription());
    }

    @Test
    public void returnFalseWhenUpdatingRssNotExist() throws Exception {
        final Rss newRss = new Rss();
        newRss.setUrl(NOT_USE_RSS_URL);

        assertTrue(mDatabase.putRssIfSameUrlNotExist(newRss));
        assertFalse(mDatabase.updateRssWithSameUrl(mRss));
    }

    @Test
    public void removeRss() throws Exception {
        assertTrue(mDatabase.putRssIfSameUrlNotExist(mRss));

        mDatabase.removeRssById(mRss.getId());

        assertTrue(mDatabase.getAllRss().isEmpty());
    }

    @Test
    public void returnFalseAfterPutRssWithSameUrl() throws Exception {
        mDatabase.putRssIfSameUrlNotExist(mRss);

        final Rss rssWithSameUrl = new Rss();
        rssWithSameUrl.setUrl(mRss.getUrl());

        assertFalse(mDatabase.putRssIfSameUrlNotExist(rssWithSameUrl));
    }

    @Test
    public void updateRssArticlesIdAfterPut() throws Exception {
        mDatabase.putRssIfSameUrlNotExist(mRss);

        for (final Article article : mRss.getArticles()) {
            assertTrue(article.getRssId() == mRss.getId());
        }
    }

    @Test
    public void updateImageBytesIfArticleExist() throws Exception {
        final byte[] newImage = {1};

        assertNotEquals(newImage.length, IMAGE.length);

        mDatabase.putRssIfSameUrlNotExist(mRss);

        for (final Article article : mRss.getArticles()) {
            article.setImage(IMAGE);

            Long id = article.getId();
            mDatabase.updateArticleImage(id, newImage);

            final Article articleFromDatabase = mDatabase.getArticleById(id);

            assertNotNull(articleFromDatabase);
            assertEquals(articleFromDatabase.getImage().length, newImage.length);
        }
    }

    @Test
    public void returnRssWithSameArticles() throws Exception {
        final HashSet<Article> articles = new HashSet<>(mRss.getArticles());

        mDatabase.putRssIfSameUrlNotExist(mRss);
        final Rss rssFromDatabase = mDatabase.getRssById(mRss.getId());

        assertNotNull(rssFromDatabase);

        for (final Article articleFromDatabase : rssFromDatabase.getArticles()) {
            assertTrue(articles.contains(articleFromDatabase));
        }
    }

    @Test
    public void canSaveImageBytesArray() throws Exception {
        for (final Article article : mRss.getArticles()) {
            article.setImage(IMAGE);
        }

        mDatabase.putRssIfSameUrlNotExist(mRss);
        final Rss rssFromDb = mDatabase.getRssById(mRss.getId());

        assertNotNull(rssFromDb);

        for (final Article article : rssFromDb.getArticles()) {
            assertNotNull(article.getImage());
        }
    }

    @Test
    public void returnAllRss() throws Exception {
        mDatabase.putRssIfSameUrlNotExist(mRss);

        assertEquals(mDatabase.getAllRss().size(), 1);
    }

    @NonNull
    private Article createArticle(int number) throws Exception {
        return new Article(
                String.format(ARTICLE_TITLE, number),
                String.format(ARTICLE_DESCRIPTION, number),
                String.format(ARTICLE_ORIGIN, number));
    }
}

File: app/src/test/java/ru/iandreyshev/parserrss/models/repository/DatabaseTest.kt
package ru.iandreyshev.parserrss.models.repository

import junit.framework.Assert

import junit.framework.TestCase

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

import org.junit.Assert.*

import java.io.File
import java.util.ArrayList
import java.util.HashSet

@RunWith(RobolectricTestRunner::class)
class DatabaseTest : TestCase() {

    companion object {
        private const val RSS_TITLE = "TITLE"
        private const val RSS_ORIGIN = "ORIGIN"
        private const val RSS_URL = "URL"
        private const val NOT_USE_RSS_URL = "NOT_USE_URL"

        private const val ARTICLES_COUNT = 25
        private const val ARTICLE_TITLE = "TITLE %s"
        private const val ARTICLE_DESCRIPTION = "DESCRIPTION %s"
        private const val ARTICLE_ORIGIN = "ORIGIN %s"
    }

    private lateinit var database: Database
    private lateinit var rss: Rss

    @Before
    @Throws(Exception::class)
    fun setup() {
        val tempFile = File.createTempFile("object-store-test", "")

        assertTrue(tempFile.delete())

        database = Database(MyObjectBox.builder().directory(tempFile).build())

        rss = Rss(
                title = RSS_TITLE,
                origin = RSS_ORIGIN,
                url = RSS_URL)
        val articleList = ArrayList<Article>()

        repeat(ARTICLES_COUNT) {
            articleList.add(createArticle(it))
        }

        rss.articles = articleList
    }

    @Test
    fun noThrowExceptionIfTryToGetItemByInvalidId() {
        try {
            Database.INVALID_IDS.forEach {
                database.getArticleById(it)
                database.getRssById(it)
                database.removeRssById(it)
            }
        } catch (ex: Exception) {
            Assert.fail()
        }
    }

    @Test
    fun returnNullIfGetRssByInvalidId() {
        Assert.assertNull(database.getRssById(-2))
    }

    @Test
    @Throws(Exception::class)
    fun returnTrueIfRssWithSameUrlExist() {
        database.putRssIfSameUrlNotExist(rss)

        Assert.assertTrue(database.isRssWithUrlExist(RSS_URL))
    }

    @Test
    @Throws(Exception::class)
    fun returnFalseIfRssWithSameUrlNotExist() {
        database.putRssIfSameUrlNotExist(rss)

        Assert.assertFalse(database.isRssWithUrlExist(NOT_USE_RSS_URL))
    }

    @Test
    @Throws(Exception::class)
    fun returnArticlesById() {
        database.putRssIfSameUrlNotExist(rss)

        for ((id) in rss.articles) {
            Assert.assertNotNull(database.getArticleById(id))
        }
    }

    @Test
    @Throws(Exception::class)
    fun updateRssIdAfterPut() {
        database.putRssIfSameUrlNotExist(rss)

        Assert.assertEquals(database.getRssById(rss.id), rss)
    }

    @Test
    @Throws(Exception::class)
    fun returnTrueWhenUpdatingRssExist() {
        val newRss = Rss(title = "Title", origin = "Origin")
        newRss.description = "Description"
        newRss.url = rss.url

        assertNotEquals(newRss.title, rss.title)
        assertNotEquals(newRss.description, rss.description)
        assertNotEquals(newRss.articles.size.toLong(), rss.articles.size.toLong())

        Assert.assertEquals(newRss.url, rss.url)

        Assert.assertTrue(database.putRssIfSameUrlNotExist(newRss))
        Assert.assertTrue(database.updateRssWithSameUrl(rss))

        assertNotEquals(newRss.title, rss.title)
        assertNotEquals(newRss.description, rss.description)
    }

    @Test
    @Throws(Exception::class)
    fun returnFalseWhenUpdatingNotExistRss() {
        val newRss = Rss(title = "", origin = "", url = NOT_USE_RSS_URL)

        Assert.assertTrue(database.putRssIfSameUrlNotExist(newRss))
        Assert.assertFalse(database.updateRssWithSameUrl(rss))
    }

    @Test
    @Throws(Exception::class)
    fun removeRss() {
        assertTrue(database.putRssIfSameUrlNotExist(rss))

        database.removeRssById(rss.id)

        assertNull(database.getRssById(rss.id))
    }

    @Test
    @Throws(Exception::class)
    fun returnFalseAfterPutRssWithSameUrl() {
        database.putRssIfSameUrlNotExist(rss)

        val rssWithSameUrl = Rss(title = "", origin = "", url = RSS_URL)
        rssWithSameUrl.url = rss.url

        Assert.assertFalse(database.putRssIfSameUrlNotExist(rssWithSameUrl))
    }

    @Test
    @Throws(Exception::class)
    fun updateRssArticlesIdAfterPut() {
        database.putRssIfSameUrlNotExist(rss)

        for ((_, rssId) in rss.articles) {
            Assert.assertTrue(rssId == rss.id)
        }
    }

    @Test
    @Throws(Exception::class)
    fun returnRssWithSameArticles() {
        val articles = HashSet(rss.articles)

        database.putRssIfSameUrlNotExist(rss)
        val rssFromDatabase = database.getRssById(rss.id)

        if (rssFromDatabase == null) {
            fail()
            return
        }

        for (articleFromDatabase in rssFromDatabase.articles) {
            Assert.assertTrue(articles.contains(articleFromDatabase))
        }
    }

    private fun createArticle(number: Int): Article {
        return Article(
                title = ARTICLE_TITLE.format(number),
                description = ARTICLE_DESCRIPTION.format(number),
                originUrl = ARTICLE_ORIGIN.format(number))
    }
}